### ISSUE TITLE
Wrong type inferred after value assignment

### ISSUE THREAD

**Issue by danpascu** [microsoft/pyright#8171]
With this code:

```py
from typing import reveal_type


class B0:
    attr: str | None = None

    def __init_subclass__(cls, attr: str | None = None, **kw: object) -> None:
        super().__init_subclass__(**kw)
        if cls.attr is None:
            cls.attr = attr


class D1(B0):
    attr = 'd1'


class D2(B0, attr='d2'):
    pass


reveal_type(B0.attr)
reveal_type(D1.attr)
reveal_type(D2.attr)

```

pyright reports this:

```
  t1.py:22:13 - information: Type of "B0.attr" is "str | None"
  t1.py:23:13 - information: Type of "D1.attr" is "str | None"
  t1.py:24:13 - information: Type of "D2.attr" is "str | None"
```

mypy reports this:

```
t1.py:22: note: Revealed type is "Union[builtins.str, None]"
t1.py:23: note: Revealed type is "builtins.str"
t1.py:24: note: Revealed type is "Union[builtins.str, None]"
```

while runtime reports this:

```
Runtime type is 'NoneType'
Runtime type is 'str'
Runtime type is 'str'
```

Now I understand that static type analyzers don't necessarily analyze what a runtime function like `__init_subclass__(...)` does and how it can affect those class attributes, so I'm not surprised about the results on `D2`, but I'm surprised at  `D1.attr` being inferred as `str | None` by pyright after I explicitly assign `attr = 'd1'` for `D1`.

I have to manually add a type annotation (`attr: str = 'd1'`) to make it accept that is a `str`.

This is with the command line pyright version 1.1.366

**Comment by erictraut** [comment]
This is intended behavior, not a bug. 

Pyright always honors explicit type declarations over simple assignments. Mypy does not.

In your code sample, you have declared the type of `attr` to be `str | None` in `B0`. You have assigned it a value in child class `D1` and `D2`, but its declared type is still `str | None`.

In fact, redeclaring its type to `str` would be a type violation because the attribute is mutable, and mutable attributes are invariant — i.e. their types cannot be redefined (even covariantly) in subclasses. So mypy's behavior here is wrong and unsafe from a type perspective.

**Comment by danpascu** [comment]
But this leads to an unexpected behavior (after I assign a string to the class attribute and I try to call a method on it, it gets flagged because None doesn't support it) and is also inconsistent between class attributes and local variables that don't get the same treatment and they don't show an error:

```py
from typing import reveal_type

class B:
  a: str | None = None

class D(B):
  a = 'ABC'

reveal_type(D.a)  # Type of "D.a" is "str | None"
b = D.a.lower()   # error: "lower" is not a known attribute of "None"
c: str = D.a      # error: Expression of type "str | None" is incompatible with declared type "str"


def f():
  a: str | None
  a = 'ABC'
  reveal_type(a)  # Type of "a" is "Literal['ABC']"
  b = a.lower()   # no error
  c: str = a      # no error
```

It's unexpected that I have to use a cast or use an assert for `D.a` right after it was assigned a string, but it works fine for local variables. The inconsistency also extends to assigning to a different variable of type `str` (c in the example above).

So why is a local variable treated differently and how is one supposed to deal with the class attribute?
Do I need to add a cast or assert every place I try to work with the class attribute?

Funny enough, it works like this:

```py
class D(B):
  pass

D.a = 'ABC'

reveal_type(D.a)  # Type of "D.a" is "Literal['ABC']"
b = D.a.lower()   # no error
c: str = D.a      # no error
```

But no one wants to assign their class attributes outside the class like that.

**Comment by erictraut** [comment]
You have to keep in mind the distinction between a "declared type" and a "narrowed type". The "declared type" is provided by a type annotation and serves as an enforced upper bound for values that can be assigned to the symbol. It's possible to assign values with a narrower type, but you cannot assign a value with a wider type without triggering a type violation. When you assign a value with a narrower type, pyright (and other static type checkers) narrow the type of that symbol locally within that execution context. Locally-narrowed types do not extend beyond an execution context (e.g. a function or method body).

In your last code sample above, the expression `D.a` is narrowed to `Literal['ABC']` locally within the module (global) execution context, so you can safely call `D.a.lower()` on this value.

In your previous examples, you were assigning a `ABC` to `D.a` in a different execution context. Outside of that execution context, it is assumed to have its declared type (`str | None`) unless/until it is assigned a new value with in that execution context.

For more details, I recommend reading [this documentation](https://microsoft.github.io/pyright/#/type-concepts).

**Comment by danpascu** [comment]
Ok, that makes sense.

But it's still unclear to me what is the recommended way of dealing with this. The usual use case for attributes declared like that is to set them to None on the base class as a placeholder because a value is not yet available and the base class is virtually an abstract class that is not usable as is, but later a subclass is supposed to fill in those attribute with "real" values that are not None at which point it becomes a concrete class that is usable. That attribute is not supposed to be mutated after that (I understand that the typing system cannot know this).

Is there a recommended approach to this programming pattern that also makes sense with static typing, or do I need to use assert/cast or type: ignore for these?

**Comment by erictraut** [comment]
If you want to write type-safe code, there are two common approaches:
1. Don't use `None` for attribute values. Make sure they're always initialized to some valid value other than `None` even in the base base.
2. Insert `assert` statements in cases where you are assuming that the value isn't `None`. This documents your assumption (for both human readers and static analysis tools) and validates the assumption at runtime.
